### PR TITLE
Support proxying devices with isochronous IN endpoints

### LINKED ISSUE #13: Fix support for proxying devices that switch interface altsettings
Even though #6 added basic support for devices that switch between interface altsettings, it might not work properly in certain cases: the endpoint handling threads might fail to be joined, as they might be blocked on libusb or Raw Gadget transfers.

Interrupting Raw Gadget transfers is simple: just send any signal to the endpoint threads via `pthread_kill` and add a no-op handler for this signal (but perhaps a signal other than `SIGINT`, as USB Proxy already uses this signal internally).

~~However, interrupting the libusb transfers is a problem.~~

~~A simple solution would be use the `timeout` argument of the libusb sync I/O API, but unfortunately libusb has [a bug](https://github.com/libusb/libusb/issues/1333) that makes sync I/O ignore the `timeout` argument.~~

~~We need to either switch to using async libusb I/O or figure out another way to interrupt libusb transfers (without fully resetting the device, as done in #14).~~

Ah, actually libusb timeouts work as expected, it's just that `receive_data` zeroes out `timeout`.

### PR DESCRIPTION
**Pull request by xairy**
With these changes, I'm able to proxy a few Logitech USB web cameras at a low resolution (160x120 pixels) via dwc3.

Proxying with a higher resolution fails to produce a picture on the host: I think because the transfer rate is too low and all the frames get dropped.

I have not tried other UDCs besides dwc3 yet.

Also fixes #13.

### CODE REVIEW

**Comment by xairy** [comment]
Tested proxying with dwc2 on Raspberry Pi 4: the initial enumeration works, but attempting to stream a picture fails with `ioctl(USB_RAW_IOCTL_EP_ENABLE): Cannot allocate memory` and `dwc2 fe980000.usb: dwc2_hsotg_ep_enable: No suitable fifo found`.

I think this happens due to the hardware limitations (no FIFOs for the used packet size?).

File: proxy.cpp [hunk 1]
```diff
@@ -572,20 +615,50 @@ void ep0_loop(int fd) {
...
+							printf("ep0: request acked\n");
+					}
+					else {
+						// Stall the request.
+						usb_raw_ep0_stall(fd);
+						continue;
+					}
+				}
+				else {
+					// For non-0-length requests, we have to trieve the data and
```
Comment (AristoChen) [inline proxy.cpp:641]: I guess you mean `retrieve` instead of `trieve`?

**Comment by AristoChen** [comment]
Hi @xairy 

Thanks for the PR! Glad to see ISOC is supported now!

Overall the PR looks good to me, I will find devices to test before merging, thanks again!

File: proxy.cpp (hunk 1 above)
Comment (xairy) [reply proxy.cpp]: Right, fixed and also clarified the comment.

**Comment by xairy** [comment]
You're welcome!

I recommend using a DWC3-based device for testing, if you have one. Raspberry Pi with its dwc2, unfortunately, doesn't work. Other UDCs I haven't tested.

**Comment by AristoChen** [comment]
I found this [document](https://www.kernel.org/doc/html/v5.0/driver-api/usb/dwc3.html) listing some vendors that are using DWC3, I will try if I can get any of them. In the mean time, May I know the model of the device that you used for testing? thanks!

**Comment by xairy** [comment]
I use a ThinkPad X1 Carbon 6th Gen laptop, but getting DWC3 enabled there is tricky: https://xairy.io/articles/thinkpad-xdci. I believe that ROCK Pi uses DWC3, but I haven't tried it myself.

**Comment by AristoChen** [comment]
Using a laptop with DWC3 sounds cool! but unfortunately a bit too expensive 😓.

I got a rockpi 4c+ recently, but it doesn't looks like the OTG port works out of the box, I will find some time to see how to fix it

**Comment by AristoChen** [comment]
I borrowed an NXP i.MX8MP recently, and it also uses `dwc3`, I will find some time to get a USB camera or something else to test this PR :)

**Comment by AristoChen** [comment]
I am able to proxy the USB webcam with 1920*1080 resolution! 

Just out of curiosity, did you run the usb-proxy in a serial console environemt or a ssh environment?

In serial console environment, I am not able to proxy high resolution(I think it is because of the printed messages are too much), while using ssh works a lot better

**Comment by xairy** [comment]
That's awesome!

I was running the proxy locally in a Terminal. Perhaps, the DWC3 controller in my laptop is just slow.

**Comment by AristoChen** [comment]
I see, I am merging the PR now, thanks again for this awesome PR! Glad to see ISOC is now supported!

**Pull request merged**